app/digitizingcontroller.cpp

```diff
@@ -193,3 +193,23 @@ QgsQuickFeatureLayerPair DigitizingController::lineFeature()
 
   return QgsQuickFeatureLayerPair(f, featureLayerPair().layer());
 }
+
+QgsPoint DigitizingController::mapCoords(QgsQuickFeatureLayerPair pair)
+{
+    if ( !pair.layer() )
+      return QgsPoint();
+
+    QgsPointXY res = mMapSettings->mapSettings().layerToMapCoordinates(pair.layer(), QgsPoint(pair.feature().geometry().asPoint()));
+    return QgsPoint(res);
+}
+
+QgsQuickFeatureLayerPair DigitizingController::changePointGeometry(QgsQuickFeatureLayerPair pair, QgsPoint point)
+{
+    QgsPointXY layerPoint = mMapSettings->mapSettings().mapToLayerCoordinates(pair.layer(), QgsPointXY(point.x(), point.y()));
+    QgsPoint* mapPoint = new QgsPoint( layerPoint );
+    fixZ(mapPoint);
+    QgsGeometry geom( mapPoint );
+
+    pair.featureRef().setGeometry(geom);
+    return pair;
```

Comment:
Why do we return the feature pair here if we get it as an argument and it is not modified in the function body?

Reply:
The pair's feature geometry is modified above. In https://github.com/lutraconsulting/input/pull/182/files#diff-dffcb24829ceef0df628841dea895e27R86, featurePanel.feature is given as the first parameter - its a recent/current feature pair. This modified pair is set to featurePanel.feature and save fn is called. It wouldnt work without assigning modified feature pair.

Reply:
Ahh right I thought it was just a reference...